sdks/python/apache_beam/io/sdf_restriction_provider.py

```diff
@@ -0,0 +1,56 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.io.restriction_trackers import SDFBoundedSourceRestrictionTracker
+from apache_beam.transforms.core import RestrictionProvider
+
+__all__ = ['SDFBoundedSourceRestrictionProvider']
+
+
+class SDFBoundedSourceRestrictionProvider(RestrictionProvider):
+  """A `RestrictionProvider` that is used by SDF for `BoundedSource`."""
+
+  def __init__(self, source, desired_chunk_size=None):
+    self._source = source
+    self._desired_chunk_size = desired_chunk_size
+    # The size cannot be calculated directly by end_pos - start_pos since the
+    # position may not be numeric(e.g., key space position).
+    self._restriction_size_map = {}
+
+  def initial_restriction(self, element):
+    # Get initial range_tracker from source
+    range_tracker = self._source.get_range_tracker(None, None)
+    return (range_tracker.start_position(), range_tracker.stop_position())
+
+  def create_tracker(self, restriction):
+    return SDFBoundedSourceRestrictionTracker(
+        self._source.get_range_tracker(restriction[0],
+                                       restriction[1]))
+
+  def split(self, element, restriction):
+    # Invoke source.split to get initial splitting results.
+    source_bundles = self._source.split(self._desired_chunk_size)
+    for source_bundle in source_bundles:
+      self._restriction_size_map[(source_bundle.start_position,
+                                  source_bundle.stop_position)]\
+        = source_bundle.weight
+      yield (source_bundle.start_position, source_bundle.stop_position)
```

Comment:
I don't believe you can assume that the source you have as the element equal to the source being returned within SourceBundle.

If you take a look at the file based source, you can see that the SourceBundle may return a ConcatSource or a _SingleFileSource.

I believe you'll want to make the "restriction" a SourceBundle and use the source from the SourceBundle.

The initial restriction could return the SourceBundle(None, source, None, None). Conveniently, this would allow you to return source_bundle.weight and fall back to source_bundle.source.estimate_size() if source_bundle.weight is None.

Reply:
I don't think I make this assumption here. Taking ```FileBaseSource``` as an example, ```FileBaseSource.split()``` returns SourceBundle constructed based on ```_SingleFileSource```:
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/filebasedsource.py#L271 .

As for falling back to source_bundle.source.estimate_size(), actually I don't think it would work at least for ```_SingleFileSource```. If looking into constructing SourceBundle in split: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/filebasedsource.py#L294, the stop_pos will be set to INF, which causes source_bundle.source.estimate_size()(https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/filebasedsource.py#L302) return inaccurate size.

Reply:
The `SDFBoundedSourceRestrictionProvider` is initialized with the source object that is passed into `Read`. After splitting, the `SDFBoundedSourceRestrictionProvider` creates a tracker using that source from `Read` and not the source contained inside the `SourceBundle`.

Good point about not relying on source_bundle.source.estimate_size() and only using the weight if it is provided.